File: apps/dashboard/src/lib/perf.ts
/**
 * ADDED (ISO 25010 performance-efficiency evidence, docs/20): in-session
 * response-time samples with p50/p95 logged to the console after each
 * measurement. Open devtools and read the `[perf]` lines — no backend, no
 * third-party telemetry (privacy stance, docs/12).
 */
const samples = new Map<string, number[]>();

/** Nearest-rank percentile; expects a sorted ascending array. */
export function percentile(sortedMs: number[], p: number): number {
  if (sortedMs.length === 0) return 0;
  const rank = Math.ceil((p / 100) * sortedMs.length);
  return sortedMs[Math.min(rank, sortedMs.length) - 1];
}

export async function timed<T>(metric: string, fn: () => Promise<T>): Promise<T> {
  const startedAt = performance.now();
  try {
    return await fn();
  } finally {
    const ms = Math.round(performance.now() - startedAt);
    const list = samples.get(metric) ?? [];
    list.push(ms);
    samples.set(metric, list);
    const sorted = [...list].sort((a, b) => a - b);
    console.log(
      `[perf] ${metric} ${ms}ms (session n=${sorted.length} ` +
        `p50=${percentile(sorted, 50)}ms p95=${percentile(sorted, 95)}ms)`,
    );
  }
}

File: apps/mobile/lib/util/perf.dart
import 'package:flutter/foundation.dart';

/// ADDED (ISO 25010 performance-efficiency evidence, docs/20): in-session
/// response-time samples with p50/p95 in the debug log. Read the `[perf]`
/// lines while driving the app — no backend, no third-party telemetry
/// (privacy stance, docs/12).
final Map<String, List<int>> _samples = {};

/// Nearest-rank percentile; expects a sorted ascending list.
int percentileMs(List<int> sortedMs, int p) {
  if (sortedMs.isEmpty) return 0;
  final rank = (p / 100 * sortedMs.length).ceil();
  return sortedMs[rank.clamp(1, sortedMs.length) - 1];
}

Future<T> timed<T>(String metric, Future<T> Function() fn) async {
  final stopwatch = Stopwatch()..start();
  try {
    return await fn();
  } finally {
    stopwatch.stop();
    final list = _samples.putIfAbsent(metric, () => []);
    list.add(stopwatch.elapsedMilliseconds);
    final sorted = [...list]..sort();
    debugPrint('[perf] $metric ${stopwatch.elapsedMilliseconds}ms '
        '(session n=${sorted.length} p50=${percentileMs(sorted, 50)}ms '
        'p95=${percentileMs(sorted, 95)}ms)');
  }
}
